### PR TITLE
[MRG+1] better input validation on `fill_between`

### LINKED ISSUE #7510: better input validation on `fill_between`
Hi,

So I am running on Python 3.5.2 and Matplotlib 1.5.3 and everything is installed via OS packet manager on Arch Linux. Here's the code I am having issue with:

```python
from matplotlib.pyplot import figure
from numpy import linspace

fig = figure()
ax = fig.gca()

x = linspace(0, 1).reshape(50, 1)

ax.fill_between(x, x, -x)  # That doesn't work!
ax.fill_between(x[:, 0], x[:, 0], -x[:, 0])  # And that ugliness does!
```
I've already checked that this is due to `where=None` setting which constructrs `where` variable via `len` function. This seems inadequate in many `numpy` scenarios as occasionally arrays will pop up with or without second dimension. If they do have a second dimension though the `where=None` will force the error below.

```
Traceback (most recent call last):
  File "fill_between_test.py", line 11, in <module>
    ax.fill_between(x, x, -x)  # That doesn't work!
  File "/usr/lib/python3.5/site-packages/matplotlib/__init__.py", line 1819, in inner
    return func(ax, *args, **kwargs)
  File "/usr/lib/python3.5/site-packages/matplotlib/axes/_axes.py", line 4596, in fill_between
    raise ValueError("Argument dimensions are incompatible")
ValueError: Argument dimensions are incompatible
```
Please advise.

### PR DESCRIPTION
**Pull request by samsontmr**
Fixes #7510

### CODE REVIEW

File: lib/matplotlib/tests/test_axes.py [hunk 1]
```diff
@@ -725,6 +725,42 @@ def test_polycollection_joinstyle():
     ax.set_ybound(0, 3)
 
 
+@raises(ValueError)
```
Comment (NelleV) [inline lib/matplotlib/tests/test_axes.py]: I have no clue why this doesn't fail for all the builds.
Raises is not defined. I think you need to use the "new" pytest way of doing this:

```python
def test_fill between_2d_x_input():
    …
    with pytest.raises(ValueError):
         ax.plot(…)
```

**Comment by anntzer** [comment]
I would put the check much lower, after conversion to arrays (`x = ma.masked_invalid(self.convert_xunits(x))` and so on).  Then you can just check for `x.ndim` and so on.

File: lib/matplotlib/axes/_axes.py [hunk 2]
```diff
@@ -4774,6 +4775,11 @@ def fill_between(self, x, y1, y2=0, where=None, interpolate=False,
         y1 = ma.masked_invalid(self.convert_yunits(y1))
         y2 = ma.masked_invalid(self.convert_yunits(y2))
 
+        for array in [('x', x), ('y1', y1), ('y2', y2)]:
```
Comment (anntzer) [inline lib/matplotlib/axes/_axes.py]: for name, array in ...

**Comment by samsontmr** [comment]
Done! @anntzer @NelleV

File: lib/matplotlib/axes/_axes.py [hunk 3]
```diff
@@ -4774,6 +4775,11 @@ def fill_between(self, x, y1, y2=0, where=None, interpolate=False,
         y1 = ma.masked_invalid(self.convert_yunits(y1))
         y2 = ma.masked_invalid(self.convert_yunits(y2))
 
+        for name, array in [('x', x), ('y1', y1), ('y2', y2)]:
+            if array.ndim > 1:
+                raise ValueError('Input passed into argument "' + name +
```
Comment (anntzer) [inline lib/matplotlib/axes/_axes.py]: `"Input passed into argument {!r}".format(name)` (or `"... %r" % name` if you're really old-styled).
Same fixes need to be applied below.

File: lib/matplotlib/axes/_axes.py [hunk 4]
```diff
@@ -4774,6 +4775,11 @@ def fill_between(self, x, y1, y2=0, where=None, interpolate=False,
         y1 = ma.masked_invalid(self.convert_yunits(y1))
         y2 = ma.masked_invalid(self.convert_yunits(y2))
 
+        for name, array in [('x', x), ('y1', y1), ('y2', y2)]:
```
Comment (NelleV) [inline lib/matplotlib/axes/_axes.py]: :+1:

File: lib/matplotlib/axes/_axes.py [hunk 5]
```diff
@@ -4934,6 +4941,11 @@ def fill_betweenx(self, y, x1, x2=0, where=None,
         x1 = ma.masked_invalid(self.convert_xunits(x1))
         x2 = ma.masked_invalid(self.convert_xunits(x2))
 
+        for array in [('x', x), ('y1', y1), ('y2', y2)]:
```
Comment (NelleV) [inline lib/matplotlib/axes/_axes.py]: I think there is a slight mistake here: the arguments are "y", "x1" and "x2".
The documentation and examples don't run.

Comment (NelleV) [inline lib/matplotlib/axes/_axes.py]: Any reason not to use `name, array` as above?

Comment (samsontmr) [inline lib/matplotlib/axes/_axes.py]: Good catch, thanks! Will change it now

**Comment by samsontmr** [comment]
Added @cleanup to the top of the tests. I suspect that was what caused some of the build jobs to fail

**Comment by anntzer** [comment]
Thanks!

**Pull request merged**